src/Npgsql/Internal/Converters/RecordObjectArrayConverter.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.Internal.Descriptors;
+using Npgsql.PostgresTypes;
+using Npgsql.Properties;
+
+namespace Npgsql.Internal.Converters;
+
+public class RecordObjectArrayConverter : PgStreamingConverter<object[]>
+{
+    readonly PgSerializerOptions _serializerOptions;
+
+    public RecordObjectArrayConverter(PgSerializerOptions serializerOptions)
+        => _serializerOptions = serializerOptions;
+
+    public override object[] Read(PgReader reader)
+        => Read(async: false, reader, CancellationToken.None).GetAwaiter().GetResult();
+
+    public override ValueTask<object[]> ReadAsync(PgReader reader, CancellationToken cancellationToken = default)
+        => Read(async: true, reader, cancellationToken);
+
+    async ValueTask<object[]> Read(bool async, PgReader reader, CancellationToken cancellationToken)
+    {
+        if (reader.ShouldBuffer(sizeof(int)))
+            await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+        var fieldCount = reader.ReadInt32();
+        var result = new object[fieldCount];
+
+        for (var i = 0; i < fieldCount; i++)
+        {
+            if (reader.ShouldBuffer(sizeof(uint) + sizeof(int)))
+                await reader.BufferData(async, sizeof(uint) + sizeof(int), cancellationToken).ConfigureAwait(false);
+
+            var typeOID = reader.ReadUInt32();
+            var fieldLen = reader.ReadInt32();
+
+            // Note that we leave .NET nulls in the object array rather than DBNull.
+            if (fieldLen == -1)
+                continue;
+
+            var postgresType =
+                _serializerOptions.TypeCatalog.GetPgType((Oid)typeOID).GetRepresentationalType()
+                ?? UnknownBackendType.Instance;
```

Comment:
This will bomb very nastily with an error "A PostgreSQL type with the oid '0' was not found in the database", you probably want to just throw the reading not supported exception if we can't find the type.

Reply:
Take a look at its handling here https://github.com/npgsql/npgsql/blob/2376963d884f537cf923294c12a183f37358884d/src/Npgsql/BackendMessages/RowDescriptionMessage.cs#L390-L397

Reply:
Done, take a look.